src/MySqlConnector/MySqlClient/MySqlDataReader.cs

```diff
@@ -555,7 +579,32 @@ public override object GetValue(int ordinal)
 			}
 		}
 
-		public override IEnumerator GetEnumerator() => new DbEnumerator(this, closeReader: false);
+	    public override T GetFieldValue<T>(int ordinal)
+	    {
+	        try
+	        {
+	            // try normal casting
+	            return base.GetFieldValue<T>(ordinal);
+	        }
+	        catch (InvalidCastException e)
+	        {
+	            try
+	            {
+	                // try casting using reflection; needed for json
+	                var dataParam = Expression.Parameter(typeof(byte[]), "data");
+	                var body = Expression.Block(Expression.Convert(dataParam, typeof(T)));
+	                var run = Expression.Lambda(body, dataParam).Compile();
+	                return (T)run.DynamicInvoke(GetValue(ordinal));
```

Comment:
I'd prefer to move this up to Pomelo.EntityFrameworkCore.MySql, which has the need to create `JsonObject<T>` objects. 

See proposed implementation at https://github.com/bgrainger/Pomelo.EntityFrameworkCore.MySql/commit/0e2a88a0612f439a9414d6b997b795ec77bc5a09